### PR TITLE
Stripe integration

### PR DESCRIPTION
**Pull request by Ru-T**
## What Changed? 
* Configure Stripe checkout for purchasing books in the bookstore
* Create charges controller to mark pending purchase as purchased and charge customer
* Only show purchases that are still pending in user's cart

## Why? 
* So users can buy books and get smarter!

### CODE REVIEW

**Comment by dkniffin** [comment]
Lol, this PR looks like a game of red-light green-light.

File: app/controllers/charges_controller.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+class ChargesController < ApplicationController
+  def new
+  end
```
Comment (dkniffin) [inline app/controllers/charges_controller.rb]: You can remove this empty method.

File: app/controllers/charges_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  def create
+    @pending_purchase = PendingPurchase.find(params[:pending_purchase_id])
```
Comment (dkniffin) [inline app/controllers/charges_controller.rb]: It looks like this instance variable isn't being used in any views, so you should make it a local variable instead.

File: app/controllers/charges_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    @pending_purchase.was_purchased
+    amount = @pending_purchase.total_price * 100
```
Comment (dkniffin) [inline app/controllers/charges_controller.rb]: Just to be clear, I think you should rename this to `amount_cents`

File: app/controllers/pending_purchases_controller.rb [hunk 4]
```diff
@@ -3,7 +3,7 @@ class PendingPurchasesController < ApplicationController
   before_action :set_pending_purchase, only: [:edit, :update]
 
   def index
-    @pending_purchases = PendingPurchase.where(user: current_user)
+    @pending_purchases = PendingPurchase.pending(current_user)
```
Comment (dkniffin) [inline app/controllers/pending_purchases_controller.rb]: This works as-is, but I would restructure it a little, and do `current_user.pending_purchases.pending`. That will require changing your `pending` method to a scope, and not passing the user in.

The reason for this is that if you later need to say, find all pending PendingPurchases, you can just do `PendingPurchase.pending`.

File: app/views/pending_purchases/index.html.erb [hunk 5]
```diff
@@ -1,16 +1,32 @@
...
+      <% @pending_purchases.each do |purchase| %>
+        <tr>
+          <td> <%= purchase.user.email %> </td>
+          <td> <%= purchase.book.title %> </td>
+          <td> <%= purchase.quantity %>
+               <%= link_to 'Edit Quantity', edit_pending_purchase_path(purchase), id: "edit_quantity" %> </td>
+          <td> <%= "$#{purchase.price_at_purchase}" %> </td>
+          <td> <%= "$#{purchase.total_price}" %> </td>
+          <td><%= render "charges/charge_form", purchase: purchase %></td>
+          <%= form_for(current_user.pending_purchases.find_by_book_id(purchase.book.id),
```
Comment (dkniffin) [inline app/views/pending_purchases/index.html.erb]: I think the `find_by_<attribute>` is frowned upon (and might even be deprecated?). Instead, do `find_by(book_id: purchase.book.id)`

File: config/routes.rb [hunk 6]
```diff
@@ -5,4 +5,5 @@
   resources :books, only: [:index, :show]
   devise_for :users, controllers: { registrations: "registrations" }
   resources :pending_purchases, except: [:new]
+  resources :charges
```
Comment (dkniffin) [inline config/routes.rb]: I'm not sure, but I don't think you're using all of the endpoints (index, new, create, show, edit, update, delete). You should disable whichever ones you're not using.

File: db/migrate/20160128192031_create_pending_purchases.rb [hunk 7]
```diff
@@ -4,7 +4,7 @@ def change
       t.integer :user_id
       t.integer :book_id
       t.integer :quantity
-      t.integer :price_at_purchase
+      t.integer :price_at_purchase, scale: 2
```
Comment (dkniffin) [inline db/migrate/20160128192031_create_pending_purchases.rb]: Interesting. Never used this before. TIL

File: features/purchase.feature [hunk 8]
```diff
@@ -1,7 +1,7 @@
 Feature: Purchasing a Book
   Scenario: Quantity adjustment
     Given I am logged into the site
-    And there are some books in the database
+    And there are 100 books in the database
```
Comment (dkniffin) [inline features/purchase.feature]: :+1:

File: spec/models/pending_purchase_spec.rb [hunk 9]
```diff
@@ -16,4 +17,26 @@
       expect(FactoryGirl.build(:pending_purchase, book_id: nil)).to_not be_valid
     end
   end
+
+  describe "#total_price" do
+    it "returns the purchase price times the quantity of the item" do
+      cart_item = FactoryGirl.build(:pending_purchase, price_at_purchase: 12, quantity: 3)
```
Comment (dkniffin) [inline spec/models/pending_purchase_spec.rb]: This should be in a `let`

File: spec/models/pending_purchase_spec.rb [hunk 10]
```diff
@@ -1,7 +1,8 @@
 require 'rails_helper'
 
 RSpec.describe PendingPurchase, type: :model do
-  let(:pending_purchase) { FactoryGirl.create(:pending_purchase) }
+  let(:user) { FactoryGirl.create(:user) }
```
Comment (dkniffin) [inline spec/models/pending_purchase_spec.rb]: Protip: If you add `config.include FactoryGirl::Syntax::Methods` into the `RSpec.configure` in `spec_helper.rb`, you can drop the `FactoryGirl.` everywhere.

File: app/controllers/charges_controller.rb (hunk 1 above)
Comment (Ru-T) [inline app/controllers/charges_controller.rb]: :+1:

**Comment by dkniffin** [comment]
Looks good!

File: app/controllers/charges_controller.rb (hunk 3 above)
Comment (Ru-T) [inline app/controllers/charges_controller.rb]: Yep

File: app/controllers/charges_controller.rb (hunk 2 above)
Comment (Ru-T) [inline app/controllers/charges_controller.rb]: Good catch.

File: app/views/pending_purchases/index.html.erb (hunk 5 above)
Comment (Ru-T) [inline app/views/pending_purchases/index.html.erb]: Fixed.

File: spec/models/pending_purchase_spec.rb (hunk 9 above)
Comment (Ru-T) [inline spec/models/pending_purchase_spec.rb]: Fixed.

File: app/models/pending_purchase.rb [hunk 11]
```diff
@@ -3,4 +3,17 @@ class PendingPurchase < ActiveRecord::Base
...
+    price_at_purchase * quantity
+  end
+
+  def was_purchased
+    self.purchased = true
+    save
+  end
+
+  def self.pending
+    PendingPurchase.all.reject { |pending_purchase| pending_purchase.purchased == true }
```
Comment (dkniffin) [inline app/models/pending_purchase.rb]: Good job switching this over. Two notes about the new code:

1. Change `PendingPurchase.all.reject ...` to `PendingPurchase.where(purchased: false)`. `reject` will iterate over all records.
2. Change this to a [scope](http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html)

**Pull request merged**